### PR TITLE
[SPARK-49687][SQL] Delay sorting in `validateAndMaybeEvolveStateSchema`

### PR DESCRIPTION
**Pull request by tedyu**
### What changes were proposed in this pull request?
In `validateAndMaybeEvolveStateSchema`, existing schema and new schema are sorted by column family name.
The sorting can be delayed until `createSchemaFile` is called.
When computing `colFamiliesAddedOrRemoved`, we can use `toSet` to compare column families.

### Why are the changes needed?
This would make `validateAndMaybeEvolveStateSchema` faster.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

Existing tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by tedyu** [comment]
@eric-maynard @HeartSaVioR 
Can you review ?

**Review by ericm-db — APPROVED** [review]
LGTM, pending CI

**Comment by tedyu** [comment]
@ericm-db 
Accoding to https://github.com/tedyu/spark/runs/30175945976, tests passed.

**Comment by tedyu** [comment]
@ericm-db 
Are you waiting for a committer to review this ?

**Comment by tedyu** [comment]
@MaxGekk 
Can you take a look ?
Thanks

**Comment by tedyu** [comment]
@dongjoon-hyun 
Can you take a look ?

Thanks

**Review by MaxGekk — COMMENTED** [review]
> This would make validateAndMaybeEvolveStateSchema faster.

Can you proof this, any microbenchmarks? BTW, building a `Set` is not for free.

And the code before and after is not logically the same. The `Set` removes duplicates naturally.  @HeartSaVioR Pleae, have a look at the changes.

**Comment by tedyu** [comment]
bq. building a Set is not for free.

Sorting is an O(N log N) operation while building Set is linear in complexity.

bq. The Set removes duplicates 

In this code, there is no duplicate in the schema.

**Review by MaxGekk — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
To @tedyu , could you file a JIRA issue, @tedyu ?

**Comment by tedyu** [comment]
@dongjoon-hyun 
I have filed https://issues.apache.org/jira/browse/SPARK-49687

Thanks

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala
Comment (tedyu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:172]: Agreed.
The sorting is done when needed.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
> Agreed. The sorting is done when needed.

No, it's not addressed properly, @tedyu . 

There are multiple places of `createSchemaFile(newStateSchemaList, stateSchemaVersion)` at line 176 and 193.

**Comment by tedyu** [comment]
I have covered the other `createSchemaFile` call.
Please take a look.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
No, please sort once.

**Comment by dongjoon-hyun** [comment]
I explicitly suggest to keep it, @tedyu . :) 
> https://github.com/apache/spark/pull/48116#discussion_r1763393870

**Comment by tedyu** [comment]
Here is why I moved the `sort` call:
```
      if (stateSchemaVersion == SCHEMA_FORMAT_V3 && colFamiliesAddedOrRemoved) {
```
when `colFamiliesAddedOrRemoved` is false, the call to `createSchemaFile` would be skipped and there is no need to sort.

Effectively the sorting is done lazily.
What do you think, @dongjoon-hyun ?

**Comment by dongjoon-hyun** [comment]
Got it. it's slightly different from what the PR title and description claims because we still have `sort`.

Could you revise the PR title and description first about when we skip sorting and when we do the sort? Then, I believe I can agree with the proposal.

**Comment by tedyu** [comment]
The PR and JIRA's titles and descriptions have been updated.

Thanks for your patience.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs). Thank you, @tedyu .

**Comment by tedyu** [comment]
@dongjoon-hyun 
Test suite passed. FYI

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @tedyu .

BTW, did you change your ID? I added you to the Apache Spark contributor group in order to assigne SPARK-49687 to you.

**Comment by tedyu** [comment]
Thanks @dongjoon-hyun 
Zhihong is my Chinese name.

**Pull request closed without merging**
